[Merged by Bors] - Migration guide 0.9-0.10 (#530)
bevyengine/bevy-website: content/learn/book/migration-guides/0.9-0.10/_index.md

    - `viewport_origin` to specify offset

- Change `ScalingMode::WindowSize` to `ScalingMode::WindowSize(1.0)`

### [Changed &mut PipelineCache to &PipelineCache](https://github.com/bevyengine/bevy/pull/7598)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">Rendering</div>
</div>

- `SpecializedComputePipelines::specialize` now takes a `&PipelineCache` instead of a `&mut PipelineCache`

### [Introduce detailed_trace macro, use in TrackedRenderPass](https://github.com/bevyengine/bevy/pull/7639)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">Rendering</div>
</div>

- Some detailed bevy trace events now require the use of the cargo feature `detailed_trace` in addition to enabling `TRACE` level logging to view. Should you wish to see these logs, please compile your code with the bevy feature `detailed_trace`. Currently, the only logs that are affected are the renderer logs pertaining to `TrackedRenderPass` functions

### [added subdivisions to shape::Plane](https://github.com/bevyengine/bevy/pull/7546)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">Rendering</div>
</div>

> All the examples needed to be updated to initalize the subdivisions field.
> Also there were two tests in tests/window that need to be updated.
> 
> A user would have to update all their uses of shape::Plane to initalize the subdivisions field.

Reviewer: ```suggestion
- `shape::Plane` now takes an additional `subdivisions` parameter so users should provide it or use the new `shape::Plane::from_size`.
```
This one I had to deduce from examples, not sure if it's 100% accurate.

### [Pipelined Rendering](https://github.com/bevyengine/bevy/pull/6503)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">Rendering</div>
    <div class="migration-guide-area-tag">Tasks</div>
</div>

__App `runner` and SubApp `extract` functions are now required to be Send__

This was changed to enable pipelined rendering. If this breaks your use case please report it as these new bounds might be able to be relaxed.

### [Remove ImageMode](https://github.com/bevyengine/bevy/pull/6674)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">Rendering</div>
    <div class="migration-guide-area-tag">UI</div>
</div>

`ImageNode` never worked, if you were using it please create an issue.

### [Rename the `background_color` of 'ExtractedUiNode` to `color`](https://github.com/bevyengine/bevy/pull/7452)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">Rendering</div>
    <div class="migration-guide-area-tag">UI</div>
</div>

- The `background_color` field of `ExtractedUiNode` is now named `color`.

> ### [Make spawn_dynamic return InstanceId](https://github.com/bevyengine/bevy/pull/6663)
> 
> <div class="migration-guide-area-tags">
>     <div class="migration-guide-area-tag">Scenes</div>
> </div>
> 
> <!-- TODO no migration required, will remove it later-->
> 

Reviewer: It was labeled was `C-Breaking-Change` but doesn't seems the case.
```suggestion
```

> ### [Parallelized transform propagation](https://github.com/bevyengine/bevy/pull/4775)
> 
> <div class="migration-guide-area-tags">
>     <div class="migration-guide-area-tag">Transform</div>
> </div>
> 
> <!-- TODO no migration required, will remove it later-->
> 

Reviewer: ```suggestion
```
There is no migration guide on PR

### [Remove the `GlobalTransform::translation_mut` method](https://github.com/bevyengine/bevy/pull/7134)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">Transform</div>
    <div class="migration-guide-area-tag">Hierarchy</div>
</div>

`GlobalTransform::translation_mut` has been removed without alternative,
if you were relying on this, update the `Transform` instead. If the given entity
had children or parent, you may need to remove its parent to make its transform
independent (in which case the new `Commands::set_parent_in_place` and
`Commands::remove_parent_in_place` may be of interest)

Bevy may add in the future a way to toggle transform propagation on
an entity basis.

### [Flip UI image](https://github.com/bevyengine/bevy/pull/6292)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">UI</div>
</div>

> <!-- TODO -->

Reviewer: ```suggestion
- `UiImage` is a struct now, so use `UiImage::new(handler)` instead of `UiImage(handler)`
- `UiImage` no longer implements `Deref` and `DereftMut`, so use `&image.texture` or ``&mut image.texture` instead
```
AFAICT this one needs a migration guide, since `UiImage` is now a struct instead of a newtype, but the PR has none. This migration guide was my best guess by reading the PR.

    <div class="migration-guide-area-tag">UI</div>
</div>

TextError::ExceedMaxTextAtlases(usize)` was never thrown so if you were matching on this variant you can simply remove it.

### [Change default FocusPolicy to Pass](https://github.com/bevyengine/bevy/pull/7161)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">UI</div>
</div>

- `FocusPolicy` default has changed from `FocusPolicy::Block` to `FocusPolicy::Pass`

### [Remove VerticalAlign from TextAlignment](https://github.com/bevyengine/bevy/pull/6807)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">UI</div>
</div>

The `alignment` field of `Text` now only affects the text’s internal alignment.

__Change `TextAlignment` to TextAlignment` which is now an enum. Replace:__

- `TextAlignment::TOP_LEFT`, `TextAlignment::CENTER_LEFT`, `TextAlignment::BOTTOM_LEFT` with `TextAlignment::Left`
- `TextAlignment::TOP_CENTER`, `TextAlignment::CENTER_LEFT`, `TextAlignment::BOTTOM_CENTER` with `TextAlignment::Center`
- `TextAlignment::TOP_RIGHT`, `TextAlignment::CENTER_RIGHT`, `TextAlignment::BOTTOM_RIGHT` with `TextAlignment::Right`

__Changes for `Text2dBundle`__

> `Text2dBundle` has a new field ‘text_anchor’ that takes an `Anchor` component that controls its position relative to its transform.
> 

Reviewer 2: ```suggestion
`Text2dBundle` has a new field ‘text_anchor’ that takes an `Anchor` component that controls its position relative to its transform.

`Text2dSize` was removed. Use `TextLayoutInfo` instead.
```

The `Size::height` constructor function now sets the `width` to `Val::Auto` instead of `Val::Undefined`.

### [The `size` field of `CalculatedSize` should not be a `Size`](https://github.com/bevyengine/bevy/pull/7641)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">UI</div>
</div>

- The size field of `CalculatedSize` has been changed to a `Vec2`.

### [Windows as Entities](https://github.com/bevyengine/bevy/pull/5589)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">Windowing</div>
</div>

- Replace `WindowDescriptor` with `Window`.
  - Change `width` and `height` fields in a `WindowResolution`, either by doing

```rust
WindowResolution::new(width, height) // Explicitly
// or using From<_> for tuples for convenience
(1920., 1080.).into()
```

- Replace any `WindowCommand` code to just modify the `Window`’s fields directly  and creating/closing windows is now by spawning/despawning an entity with a `Window` component like so:

```rust
let window = commands.spawn(Window { ... }).id(); // open window
commands.entity(window).despawn(); // close window

> ```

Reviewer 3: This is missing a paragraph explaining how to replace usage of `Res<Windows>` here is an example:

----

Now that windows are components on entity, you should use `Query` instead of `Res` to access windows

**Before:**
```rust
fn count_pixels(windows: Res<Windows>) {
  let Some(primary) = windows.get_primary() else { return; };
  println!("{}", primary.width() * primary.height());
}
```
**After:**
```rust
use bevy::window::PrimaryWindow;
fn count_pixels(primary_query: Query<&Window, With<PrimaryWindow>>) {
  let Ok(primary) = primary_query.get_single() else { return; };
  println!("{}", primary.width() * primary.height());
}
```

Author: ```suggestion
### [Windows as Entities](https://github.com/bevyengine/bevy/pull/5589)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">Windowing</div>
</div>

- Replace `WindowDescriptor` with `Window`.
  - Change `width` and `height` fields in a `WindowResolution`, either by doing

```rust
WindowResolution::new(width, height) // Explicitly
// or using From<_> for tuples for convenience
(1920., 1080.).into()
```

- Replace any `WindowCommand` code to just modify the `Window`’s fields directly  and creating/closing windows is now by spawning/despawning an entity with a `Window` component like so:

```rust
let window = commands.spawn(Window { ... }).id(); // open window
commands.entity(window).despawn(); // close window
```

Now that windows are components on entity, you should use Query instead of Res to access windows

```rust
fn count_pixels(windows: Res<Windows>) { let Some(primary) = windows.get_primary() else { return; }; println!("{}", primary.width() * primary.height()); }
```

```rust
use bevy::window::PrimaryWindow; fn count_pixels(primary_query: Query<&Window, With<PrimaryWindow>>) { let Ok(primary) = primary_query.get_single() else { return; }; println!("{}", primary.width() * primary.height()); }
```

```

> ### [Allow not preventing default event behaviors on wasm](https://github.com/bevyengine/bevy/pull/7304)
> 
> <div class="migration-guide-area-tags">
>     <div class="migration-guide-area-tag">Windowing</div>
> </div>
> 
> <!-- TODO I'm not sure this needs a guide, I assume most people would be using the ..default() anyway and the ones that aren't doing that will just have a clear compile error -->
> 

Reviewer: ```suggestion
```

### [Break `CorePlugin` into `TaskPoolPlugin`, `TypeRegistrationPlugin`, `FrameCountPlugin`.](https://github.com/bevyengine/bevy/pull/7083)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">Core</div>
</div>

- `CorePlugin` broken into separate plugins.  If not using `DefaultPlugins` or `MinimalPlugins` `PluginGroup`s, the replacement for `CorePlugin` is now to add `TaskPoolPlugin`, `TypeRegistrationPlugin`, and `FrameCountPlugin` to the app.

### [Replace `RemovedComponents<T>` backing with `Events<Entity>`](https://github.com/bevyengine/bevy/pull/5680)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">ECS</div>
</div>

- Add a `mut` for `removed: RemovedComponents<T>` since we are now modifying an event reader internally.
- Iterating over removed components now requires `&mut removed_components` or `removed_components.iter()` instead of `&removed_components`.

### [Remove `ExclusiveSystemParam::apply`](https://github.com/bevyengine/bevy/pull/7489)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">ECS</div>
</div>

> _Note for maintainers: this migration guide makes more sense if it’s placed above the one for #6919._

Reviewer 4: This should be addressed. Probably should be merged into that PR's migration guide.

Author: For things like that, I prefer waiting until closer to release. Since this is generated manually, it will be really noisy to move it everytime I re-generate it.

Before:

```rust
fn parallel_system(query: Query<&MyComponent>) {
   query.par_for_each(32, |comp| {
        ...
   });
}
```

After:

```rust
fn parallel_system(query: Query<&MyComponent>) {
   query.par_iter().for_each(|comp| {
        ...
   });
}
```

### [Support piping exclusive systems](https://github.com/bevyengine/bevy/pull/7023)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">ECS</div>
</div>

Exclusive systems (systems that access `&mut World`) now support system piping, so the `ExclusiveSystemParamFunction` trait now has generics for the `In`put and `Out`put types.

```rust
// Before
fn my_generic_system<T, Param>(system_function: T)
where T: ExclusiveSystemParamFunction<Param>
{ ... }

// After
fn my_generic_system<T, In, Out, Param>(system_function: T)
where T: ExclusiveSystemParamFunction<In, Out, Param>
{ ... }
```

### [Document alignment requirements of `Ptr`, `PtrMut` and `OwningPtr`](https://github.com/bevyengine/bevy/pull/7151)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">ECS</div>
</div>

> - Safety invariants on `bevy_ptr` types’ `new` `byte_add` and `byte_offset` methods have been changed. All callers should re-audit for soundness.

Reviewer 4: Breaking changes to safety invariants seem like they could insidiously hard to find, and the consequences result in UB. Should we show some higher priority to these?

Author: Right now, the only sort is by area and then by close date. I have no issue with changing ordering, although, I think this should keep the area ordering. So in this case, keep it close to other ECS related guides. As mentioned in another comment, I'd prefer waiting until closer to release before actually moving it so I'll leave the conversation opened.

impl SystemParam for MyParam<'_, '_> {
    type State = MyParamState;
}
unsafe impl SystemParamState for MyParamState {
    fn init(world: &mut World, system_meta: &mut SystemMeta) -> Self { ... }
}
unsafe impl<'w, 's> SystemParamFetch<'w, 's> for MyParamState {
    type Item = MyParam<'w, 's>;
    fn get_param(&mut self, ...) -> Self::Item;
}
unsafe impl ReadOnlySystemParamFetch for MyParamState { }

// After (0.10)
unsafe impl SystemParam for MyParam<'_, '_> {
    type State = MyParamState;
    type Item<'w, 's> = MyParam<'w, 's>;
    fn init_state(world: &mut World, system_meta: &mut SystemMeta) -> Self::State { ... }
    fn get_param<'w, 's>(state: &mut Self::State, ...) -> Self::Item<'w, 's>;
}
unsafe impl ReadOnlySystemParam for MyParam<'_, '_> { }
```

The trait `ReadOnlySystemParamFetch` has been replaced with `ReadOnlySystemParam`.

```rust
// Before
unsafe impl ReadOnlySystemParamFetch for MyParamState {}

// After
unsafe impl ReadOnlySystemParam for MyParam<'_, '_> {}
```

### [Extend EntityLocation with TableId and TableRow](https://github.com/bevyengine/bevy/pull/6681)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">ECS</div>
</div>

> A `World` can only hold a maximum of 232 - 1 archetypes and tables now. If your use case requires more than this, please file an issue explaining your use case.

Reviewer 4: ```suggestion
A `World` can only hold a maximum of 2<sup>32</sup> - 1 archetypes and tables now. If your use case requires more than this, please file an issue explaining your use case.
```

Not sure why the superscript here didn't transfer over.

Author: Right now this is generate by parsing the section as markdown and then rewriting markdown manually not just pasting the code directly. It's worked fine before, but I'm not sure why it skipped this

+++
title = "0.9 to 0.10"
weight = 6
sort_by = "weight"
template = "book-section.html"
page_template = "book-section.html"
insert_anchor_links = "right"
[extra]
long_title = "Migration Guide: 0.9 to 0.10"
+++

Bevy relies heavily on improvements in the Rust language and compiler.
As a result, the Minimum Supported Rust Version (MSRV) is "the latest stable release" of Rust.
<div class="migration-guide">

### [bevy_reflect: Pre-parsed paths](https://github.com/bevyengine/bevy/pull/7321)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">Animation</div>
    <div class="migration-guide-area-tag">Reflection</div>
</div>

`GetPath` methods have been renamed according to the following:

- `path` -> `reflect_path`
- `path_mut` -> `reflect_path_mut`
- `get_path` -> `path`
- `get_path_mut` -> `path_mut`

### [Remove App::add_sub_app](https://github.com/bevyengine/bevy/pull/7290)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">App</div>
</div>

> `App::add_sub_app` has been removed in favor of `App::insert_sub_app`. Use `SubApp::new` and insert it via `App::add_sub_app`

Reviewer 5: ```suggestion
`App::add_sub_app` has been removed in favor of `App::insert_sub_app`. Use `SubApp::new` and insert it via `App::insert_sub_app`

```

### [Migrate engine to Schedule v3](https://github.com/bevyengine/bevy/pull/7267)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">Rendering</div>
    <div class="migration-guide-area-tag">ECS</div>
</div>

- Calls to `.label(MyLabel)` should be replaced with `.in_set(MySet)`
- Stages have been removed. Replace these with system sets, and then add command flushes using the `apply_system_buffers` exclusive system where needed.
- The `CoreStage`, `StartupStage,`RenderStage`and`AssetStage`enums have been replaced with`CoreSet`,`StartupSet, `RenderSet` and `AssetSet`. The same scheduling guarantees have been preserved.
  - Systems are no longer added to `CoreSet::Update` by default. Add systems manually if this behavior is needed, although you should consider adding your game logic systems to `CoreSchedule::FixedTimestep` instead for more reliable framerate-independent behavior.
  - Similarly, startup systems are no longer part of `StartupSet::Startup` by default. In most cases, this won’t matter to you.
  - For example, `add_system_to_stage(CoreStage::PostUpdate, my_system)` should be replaced with

>   - `add_system(my_system.in_set(CoreSet::PostUpdate)`

Reviewer 6: I think this needs to be `.in_base_set` now

Reviewer 3: @IceSentry this comment might have been overlooked. For `CoreSet`, you need to use `in_base_set`

Author: ```suggestion
  - `add_system(my_system.in_base_set(CoreSet::PostUpdate)`

```

- Fixed timesteps are now evaluated in a schedule, rather than controlled via run criteria. The `run_fixed_timestep` system runs this schedule between `CoreSet::First` and `CoreSet::PreUpdate` by default.
- Command flush points introduced by `AssetStage` have been removed. If you were relying on these, add them back manually.
- the `calculate_bounds` system, with the `CalculateBounds` label, is now in `CoreSet::Update`, rather than in `CoreSet::PostUpdate` before commands are applied. You may need to order your movement systems to occur before this system in order to avoid system order ambiguities in culling behavior.
- the `RenderLabel` `AppLabel` was renamed to `RenderApp` for clarity
- `App::add_state` now takes 0 arguments: the starting state is set based on the `Default` impl.
- Instead of creating `SystemSet` containers for systems that run in stages, simply use `.on_enter::<State::Variant>()` or its `on_exit` or `on_update` siblings.
- `SystemLabel` derives should be replaced with `SystemSet`. You will also need to add the `Debug`, `PartialEq`, `Eq`, and `Hash` traits to satisfy the new trait bounds.

> - `with_run_criteria` has been renamed to `run_if`. Run criteria have been renamed to run conditions for clarity, and should now simply return a bool.

Reviewer 6: Can we add a mention to `schedule::SchouldRun` here? Then it's easier to Ctrl+F find this change when you get the compile error.

```suggestion
- `with_run_criteria` has been renamed to `run_if`. Run criteria have been renamed to run conditions for clarity, and should now simply return a `bool` instead of `schedule::ShouldRun`.

```

- Strings can no longer be used as a `SystemLabel` or `SystemSet`. Use a type, or use the system function instead.

### [Pipelined Rendering](https://github.com/bevyengine/bevy/pull/6503)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">Rendering</div>
    <div class="migration-guide-area-tag">Tasks</div>
</div>

__App `runner` and SubApp `extract` functions are now required to be Send__

This was changed to enable pipelined rendering. If this breaks your use case please report it as these new bounds might be able to be relaxed.

### [Rename the `background_color` of 'ExtractedUiNode` to `color`](https://github.com/bevyengine/bevy/pull/7452)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">Rendering</div>
    <div class="migration-guide-area-tag">UI</div>
</div>

The `background_color` field of `ExtractedUiNode` is now named `color`.

### [Remove ImageMode](https://github.com/bevyengine/bevy/pull/6674)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">Rendering</div>
    <div class="migration-guide-area-tag">UI</div>
</div>

> `ImageNode` never worked, if you were using it please create an issue.

Reviewer 7: ```suggestion
`ImageMode` never worked, if you were using it please create an issue.

```

</div>

`ManualEventIterator` and `ManualEventIteratorWithId` are no longer `DoubleEndedIterator`s since the impls didn't work correctly, and any code using this was likely broken.

### [Rename `Tick::is_older_than` to `Tick::is_newer_than`](https://github.com/bevyengine/bevy/pull/7561)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">ECS</div>
</div>

Replace usages of `Tick::is_older_than` with `Tick::is_newer_than`.

### [Cleanup system sets called labels](https://github.com/bevyengine/bevy/pull/7678)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">ECS</div>
</div>

`PrepareAssetLabel` is now called `PrepareAssetSet`

### [Simplify generics for the `SystemParamFunction` trait](https://github.com/bevyengine/bevy/pull/7675)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">ECS</div>
</div>

For the `SystemParamFunction` trait, the type parameters `In`, `Out`, and `Param` have been turned into associated types.

```rust
// 0.9
fn my_generic_system<T, In, Out, Param, Marker>(system_function: T)
where
    T: SystemParamFunction<In, Out, Param, Marker>,

>     T: Param: SystemParam,

Reviewer 8: ```suggestion
    Param: SystemParam,
```

### [Remove the `GlobalTransform::translation_mut` method](https://github.com/bevyengine/bevy/pull/7134)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">Transform</div>
    <div class="migration-guide-area-tag">Hierarchy</div>
</div>

`GlobalTransform::translation_mut` has been removed without alternative,
if you were relying on this, update the `Transform` instead. If the given entity
had children or parent, you may need to remove its parent to make its transform
independent (in which case the new `Commands::set_parent_in_place` and
`Commands::remove_parent_in_place` may be of interest)

Bevy may add in the future a way to toggle transform propagation on
an entity basis.

### [Flip UI image](https://github.com/bevyengine/bevy/pull/6292)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">UI</div>
</div>

- `UiImage` is a struct now, so use `UiImage::new(handler)` instead of `UiImage(handler)`
- `UiImage` no longer implements `Deref` and `DereftMut`, so use `&image.texture` or ``&mut image.texture` instead

### [Remove `TextError::ExceedMaxTextAtlases(usize)` variant](https://github.com/bevyengine/bevy/pull/6796)

> <div class="migration-guide-area-tags">

Reviewer 9: Formatting is broken in the preview starting here. Is there an unclosed tag or something?

Author: It seems to work correctly for me when running it with zola locally, do you have a screenshot?

```

Replace any `WindowCommand` code to just modify the `Window`’s fields directly  and creating/closing windows is now by spawning/despawning an entity with a `Window` component like so:

```rust
let window = commands.spawn(Window { ... }).id(); // open window
commands.entity(window).despawn(); // close window
```

To get a window, you now need to use a `Query` instead of a `Res`

```rust
// 0.9
fn count_pixels(windows: Res<Windows>) {
    let Some(primary) = windows.get_primary() else {
        return;
    };
    println!("{}", primary.width() * primary.height());
}

// 0.10
fn count_pixels(primary_query: Query<&Window, With<PrimaryWindow>>) {
    let Ok(primary) = primary_query.get_single() else {
        return;
    };
    println!("{}", primary.width() * primary.height());
}
```

### [Update winit to 0.28](https://github.com/bevyengine/bevy/pull/7480)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">Windowing</div>
</div>

```rust
// 0.9
app.new()
    .add_plugins(DefaultPlugins.set(WindowPlugin {
        primary_window: Some(Window {
            always_on_top: true,
            ..default()
        }),
        ..default()
    }));

> // 0.9

Reviewer 10: ```suggestion
// 0.10
```

### [Remove dead code after #7784](https://github.com/bevyengine/bevy/pull/7875)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">Rendering</div>
</div>

Removed `SetShadowViewBindGroup`, `queue_shadow_view_bind_group()`, and `LightMeta::shadow_view_bind_group` in favor of reusing the prepass view bind group.

### [Directly extract joints into SkinnedMeshJoints](https://github.com/bevyengine/bevy/pull/6833)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">Rendering</div>
    <div class="migration-guide-area-tag">Animation</div>
</div>

`ExtractedJoints` has been removed. Read the bound bones from `SkinnedMeshJoints` instead.

### [Intepret glTF colors as linear instead of sRGB](https://github.com/bevyengine/bevy/pull/6828)

<div class="migration-guide-area-tags">
    <div class="migration-guide-area-tag">Rendering</div>
    <div class="migration-guide-area-tag">Assets</div>
</div>

No api changes are required, but it's possible that your gltf meshes look different

> 

Reviewer 11: Not sure if this will get automatically picked up, so adding it here as a suggestion, related to the item above:

```suggestion

### [Send emissive color to uniform as linear instead of sRGB](https://github.com/bevyengine/bevy/pull/7897)

- If you have previously manually specified emissive values with `Color::rgb()` and would like to retain the old visual results, you must now use `Color::rgb_linear()` instead;
- If you have previously manually specified emissive values with `Color::rgb_linear()` and would like to retain the old visual results, you'll need to apply a one-time gamma calculation to your channels manually to get the _actual_ linear RGB value: 
  - For channel values greater than `0.0031308`, use `(1.055 * value.powf(1.0 / 2.4)) - 0.055`;
  - For channel values lower than or equal to `0.0031308`, use `value * 12.92`;
- Otherwise, the results should now be more consistent with other tools/engines.
```

- the `RenderLabel` `AppLabel` was renamed to `RenderApp` for clarity
- `App::add_state` now takes 0 arguments: the starting state is set based on the `Default` impl.
- Instead of creating `SystemSet` containers for systems that run in stages, simply use `.on_enter::<State::Variant>()` or its `on_exit` or `on_update` siblings.
- `SystemLabel` derives should be replaced with `SystemSet`. You will also need to add the `Debug`, `PartialEq`, `Eq`, and `Hash` traits to satisfy the new trait bounds.
- `with_run_criteria` has been renamed to `run_if`. Run criteria have been renamed to run conditions for clarity, and should now simply return a `bool` instead of `schedule::ShouldRun`.

- States have been dramatically simplified: there is no longer a “state stack”. To queue a transition to the next state, call `NextState::set`
- Strings can no longer be used as a `SystemLabel` or `SystemSet`. Use a type, or use the system function instead.

#### Multiple fixed timesteps

> Only one fixed timestep is supported. If you were relying on this functionality, you should swap to using timers, via the on_timer(MY_PERIOD) run condition.

Reviewer 9: ```suggestion
Apps may now only have one unified fixed timestep. If you were relying on this functionality, you should swap to using timers, via the `on_timer(MY_PERIOD)` run condition.

```
